torch/distributed/algorithms/_comm_hooks/default_hooks.py

```diff
@@ -0,0 +1,126 @@
+import torch
+import torch.distributed as dist
+from torch.distributed import distributed_c10d
+
+
+class DefaultState(object):
+    r"""
+    Stores state needed to perform the default ``all_reduce`` algorithm
+    within a communication hook.
+
+    Args:
+        process_group (ProcessGroup): The process group to be used for all-reduce.
+        world_size (int): The number of workers in a process group.
+            Determined based on a ``process_group``.
+        gradient_predivide_factor (float): A factor for gradients' pre-division.
+        gradient_postdivide_factor (float): A factor for gradients' post-division.
+    """
+
+    __slots__ = [
+        "process_group",
+        "world_size",
+        "gradient_predivide_factor",
+        "gradient_postdivide_factor"
+    ]
+
+    def __init__(
+        self,
+        process_group
+    ):
+        self.process_group = process_group if process_group is not None else distributed_c10d._get_default_group()
+        self.world_size = dist.get_world_size(process_group)
+        self.gradient_predivide_factor = self._get_gradient_predivide_factor(
+            self.world_size
+        )
+        self.gradient_postdivide_factor = self.world_size / self.gradient_predivide_factor
+
+    # setting two factors `self.gradient_predivide_factor`
+    # and `self.gradient_postdivide_factor` to avoid underflow and overflow
+    def _get_gradient_predivide_factor(self, world_size: int) -> float:
+        factor: int = 1
+        while world_size % factor == 0 and world_size / factor > factor:
+            factor *= 2
+        return float(factor)
+
+class LowPrecisionState(DefaultState):
+    r"""
```

Comment:
Mention that it casts gradients back to the original parameter precision specified by `parameter_type` in the doc?